snap: Use native wayland client when possible

3v1n0 opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/microsoft/vscode/wiki/How-to-Contribute#pull-requests
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->
Electron supports wayland natively since version 12, but when running code from the snap it's not possible to use because of some issues.

Handle them here so that a native client can be used when in wayland (unless `DISABLE_WAYLAND` env variable is used), making code to have proper rendering in HiDPI screens.

This behavior can be reverted using `snap set code wayland-native=false`.

Before:
![Schermata del 2023-11-24 18-24-39](https://github.com/microsoft/vscode/assets/345675/49150d6b-b252-4914-a4d9-81ee1eb05330)

After:
![Schermata del 2023-11-24 18-25-24](https://github.com/microsoft/vscode/assets/345675/10f30002-1fb7-4139-b35f-5c45cb330ae4)

3v1n0 committed:
snap/electron-launch: Properly detect Wayland socket

Code is a classic snap, as such it has access to the whole rootfs
and by not using plugs it does not use any special path for XDG
runtime dir. So we can check the wayland socket presency by
just looking in the canonical path.

3v1n0 committed:
snap: Add dependencies to run under wayland

When running under wayland the gtk version embeded in the
snap will try to access to some xsettings provided by gnome
settings daemon, however there's no schema available and
this would read to a fatal error.

To prevent this, install the g-s-d schemas in the snap so that
the app could try to get such settings without crashing.

3v1n0 committed:
snap: Use graphics libraries from distro version matching core

The libraries shipped with code may (and in fact they do!) not be
compiled for the same glibc version of the host system, so we
instead need to use the ones that are matching the core version
we're targgeting to.

3v1n0 committed:
snap: Add patchelf to build-packages

This tool is used during build so it should be included in the build-packages list.

3v1n0 committed:
snap/electron-launch: Use ozone wayland platform under wayland

3v1n0 commented:
@microsoft-github-policy-service agree

3v1n0 committed:
snap: Make wayland-native client configurable

By default the snap native client will not be used in wayland but this can
now be changed using:
  snap set code wayland-native=true

deepak1556 reviewed (changes requested):
Thanks for the PR! starting with some high level feedback,

- For the wayland changes, I would like to avoid `auto` being the default for reasons mentioned in https://github.com/microsoft/vscode/pull/186993#pullrequestreview-1523395193. Also in favor of not just making a snap specific solution, it would be great if you can provide a [runtime argument](https://github.com/microsoft/vscode/blob/055458feab40fc308fc6d70b185068138c8cf8eb/src/vs/workbench/electron-sandbox/desktop.contribution.ts#L319-L383) to support the `ozone-platform-hint` flags which can then be used across deb, rpm packages as well.

- For the hiDPI changes, can you separate them into a different PR and provide additional context on the bug would be great.

3v1n0 commented:
> For the wayland changes, I would like to avoid auto being the default for reasons mentioned in https://github.com/microsoft/vscode/pull/186993#pullrequestreview-1523395193. Also in favor of not just making a snap specific solution, it would be great if you can provide a [runtime argument](https://github.com/microsoft/vscode/blob/055458feab40fc308fc6d70b185068138c8cf8eb/src/vs/workbench/electron-sandbox/desktop.contribution.ts#L319-L383) to support the ozone-platform-hint flags which can then be used across deb, rpm packages as well.

I see, I can definitely revert this to be opt-in.

> For the hiDPI changes, can you separate them into a different PR and provide additional context on the bug would be great.

That's the same change :), I mean, enabling wayland native client makes code to support native scaling.

3v1n0 committed:
main: Add --enable-native-wayland-client argument flag on linux

Allow to run code as native wayland client under linux, adding a clearer
flag that enables the required options.

H-G-Hristov commented:
Is there any progress on this?

cjwilsontech commented:
@deepak1556 Could you update your review? Would really like to see this merged with Wayland becoming the default in KDE 6.

3v1n0 committed:
snap/electron-launch: Use mesa drivers from SNAP to run code

In order to have proper acceleration code needs to access to the
GL drivers, however this is not happening right now as code will instead
use the host drivers, that may not work (as they may depend on
incompatible libraries).

While this is kinda of optional in X11, in wayland it's a required
dependency, not to make vscode to crash, so set the LIBGL_DRIVERS_PATH
accordingly

boukendesho commented:
@deepak1556 Sir, Could you share some updates on it?
